pages/workspaceLogs/workspaceLogs.js

```diff
@@ -0,0 +1,166 @@
+const sqldb = require('../../prairielib/lib/sql-db');
+const sqlLoader = require('../../prairielib/lib/sql-loader');
+
+const { WorkspaceLogs, WorkspaceVersionLogs } = require('./workspaceLogs.html');
+const fetch = require('node-fetch').default;
+
+const sql = sqlLoader.loadSqlEquiv(__filename);
+
+/**
+ * Given a list of workspace logs for a specific version sorted by date in
+ * ascending order, checks if the logs are considered expired.
+ *
+ * @param {any[]} workspaceLogs
+ * @returns {boolean}
+ */
+function areContainerLogsExpired(workspaceLogs) {
+  if (config.workspaceLogsExpirationDays === null) {
+    // Expiration is disabled.
+    return false;
+  }
+
+  if (workspaceLogs.length === 0) return false;
+  const firstLog = workspaceLogs[0];
+  return firstLog.date < config.workspaceLogsExpirationDays * 24 * 60 * 60 * 1000;
+}
+
+function areContainerLogsEnabled() {
+  return config.workspaceLogsS3Bucket !== null;
+}
+
+/**
+ * Loads all the logs for a given workspace version.
+ *
+ * The logs for the current running version, if any, are fetched from the
+ * workspace host directly. We also load all the logs for the given version
+ * from S3. Together, this gives us all the logs for the given version.
+ *
+ * @returns {Promise<string | null>}
+ */
+async function loadLogsForWorkspaceVersion(workspaceId, version) {
+  // Get the current workspace version.
+  const workspaceRes = await sqldb.queryOneRowAsync(sql.select_workspace, {
+    workspace_id: workspaceId,
+    version,
+  });
+  const workspace = workspaceRes.rows[0];
+
+  const logParts = [];
+
+  // Load the logs from S3. When workspaces are rebooted, we write the logs to
+  // an object before shutting down the container. This means that we may have
+  // multiple objects for each container. Load all of them. The objects are keyed
+  // such that they are sorted by date, so we can just load them in order.
+  const s3Client = new AWS.S3({ maxRetries: 3 });
+  const logItems = await s3Client
+    .listObjectsV2({
+      Bucket: config.workspaceLogsS3Bucket,
+      Prefix: `${workspaceId}/${version}/`,
+      MaxKeys: 1000,
+    })
+    .promise();
+
+  if (logItems.Contents.length > 0) {
+    // Load all parts serially to avoid hitting S3 rate limits.
+    for (const item of logItems.Contents) {
+      const res = await s3Client
+        .getObject({
+          Bucket: config.workspaceLogsS3Bucket,
+          Key: item.Key,
+        })
+        .promise();
+      logParts.push(res.Body.toString('utf-8'));
+    }
+  }
+
+  // If the current workspace version matches the requested version, we can
+  // reach out to the workspace host directly to get the remaining logs. Otherwise,
+  // they should have been flushed to S3 already.
+  //
+  // To avoid a race condition where a workspace shuts down by the time we reach
+  // out to the host, we'll fall back to attempting to load the logs from S3 if
+  // the host doesn't have any logs.
```

Comment:
I don't see anything below where we are explicitly falling back to reading from S3? It looks like we just "read everything from S3" and then "read from host". I think the code is fine like it is, but this comment was just a bit confusing to me.

Reply:
Good catch! This comment is just out of date. Originally I had it read only from the host _or_ S3, but then I realized that I always need to do both for multi-version workspaces. I'll update this comment to reflect what's actually happening in the latest code.

Reply:
Fixed in 252279f.